### ISSUE TITLE
SAP Dynpro Popups for certain Artifacts with namespace when pulling  

### ISSUE THREAD

**Issue by thorsten-wolf-neptune** [abapGit/abapGit#6314]
Hi all,

this is an issue originally posted here https://github.com/abapGit/abapgit-api-rfc/issues/7
but after discussion with @larshp it makes sense to have this here for general abapGit discussion.

When pulling certain artifacts that have a namespace the logic for deserializing them would call SAP Standard functionality that creates a dynpro popup.
For "normal" abapgit useage as of now this is fine but moving abapGit to a non sapgui solution this might be a problem there.

In any case because of these popups the https://github.com/abapGit/abapgit-api-rfc tool currently leads to a dump in these situations when calling ZABAPGIT_API_RFC_PULL via rfc (DYNPRO_SEND_IN_BACKGROUND) due to the lack of a frontend.
[dump.pdf](https://github.com/abapGit/abapgit-api-rfc/files/11757307/dump.pdf)

Here the research i did so far around it: When trying to pull a repository that contains SAP artifacts that are having a namespace. 
Then a Dynpro Popup would show  up for certain SAP artifact types:
![image](https://github.com/abapGit/abapgit-api-rfc/assets/83953375/d2aa3110-a5cb-402c-bbe6-ba8d34aca627)

The message in that dynpro is always the same (at least in our case) and is related to the namespace.
As of now i identified:
**ACID** -> Checkpoint Groups (Transaction SAAB)
![image](https://github.com/abapGit/abapgit-api-rfc/assets/83953375/7ea088d8-6a0d-4807-b9b8-fd6d380e3d10)

**ENHS** -> Enhancement Spots (Transaction SE20)
![image](https://github.com/abapGit/abapGit/assets/83953375/2723065b-f3fb-490a-8137-b4fcb77a88a4)

and **SOTR** Texts (come into play when using exception classes that are not message class based)
![image](https://github.com/abapGit/abapgit-api-rfc/assets/83953375/748feed8-b4fd-42ef-8a33-076425ecabd6)

It comes up because in the deserialize process some internal sap function modules/classes call function module **TRINT_OBJECTS_CHECK_AND_INSERT** with parameter IV_DIALOG <> space

so abapgit cannot really influence it 😞 

I see one "dirty" workaround:
That popup has two buttons: 1 "normal" ok 
![image](https://github.com/abapGit/abapgit-api-rfc/assets/83953375/d6960c11-375a-434b-aa92-e3b4da3151d8)

and 2. "Furhter for all".
![image](https://github.com/abapGit/abapgit-api-rfc/assets/83953375/9cbb8add-b220-4e42-9f50-9bbf5a63e44a)

SAP will check if the message that is about to be shown already accepted with "OK for all" with a global internal table called **gt_confirmed_messages**. If so they don't show that popup:
![image](https://github.com/abapGit/abapgit-api-rfc/assets/83953375/8a0bf8d9-109c-44b8-b23a-b5ad85a548a5)

so in my very dirty workaround i did the following: hack the global internal table with a dirty assign 😄 
At the very beginning of ZABAPGIT_API_RFC_PULL the following code works to bypass the issue:
![image](https://github.com/abapGit/abapgit-api-rfc/assets/83953375/4ffbe60f-555c-42c5-927b-6d8b503c99c4)
```abap
  " BEGIN WORKAROUND FOR POPUP "Object can only be created  in package of namespace ..."
  types: begin of gty_message,
           id type symsgid,
           ty type symsgty,
           no type symsgno,
           v1 type symsgv,
           v2 type symsgv,
           v3 type symsgv,
           v4 type symsgv,
         end of gty_message,
         gtyt_message type standard table of gty_message.

  data: ls_confirmed_message type gty_message.

  field-symbols: <gt_confirmed_messages> type gtyt_message.

  " dummy call to load the function group into memory in order to access the global gt_confirmed_messages itab.
  call function 'TR_READ_CURRENT_ACTIVITY'.

  " try to get the gt_confirmed_messages itab.
  assign ('(SAPLSTRD)GT_CONFIRMED_MESSAGES') to <gt_confirmed_messages>.
  if sy-subrc = 0.
    " got it! now simulate there has been a "Further for all" button press for the necessary messages.
    " that prevents the popup to be shown in (SAPLSTRD)raise_error_or_send_popup
    " as of now we are only aware of the message "Object can only be created  in package of namespace ..."
    ls_confirmed_message-id = 'TR'.
    ls_confirmed_message-no = '007'.
    insert ls_confirmed_message into table <gt_confirmed_messages>.
  endif.
  " END WORKAROUND FOR POPUP "Object can only be created  in package of namespace ..."
```

any ideas what else we could do to cover this problem? Dirty assigns are the worst to do i know 😃 but i am afraid that is the only solution we have for this...

**Comment by fabianlupa** [comment]
> any ideas what else we could do to cover this problem? Dirty assigns are the worst to do i know 😃 but i am afraid that is the only solution we have for this...

Here's another bad solution: Wrap the functionality in a program with a transaction code and use batch input to confirm the popup ... The dirty assign doesn't feel so bad now, does it? :P

**Comment by thorsten-wolf-neptune** [comment]
true 😄 that would be even worse.

**Comment by larshp** [comment]
@thorsten-wolf-neptune suggest adding a new method in this class and calling it from around https://github.com/abapGit/abapGit/blob/b0c213b8239495f58e6b872799848a43d9bae21c/src/objects/zcl_abapgit_objects.clas.abap#L666

**Comment by mbtools** [comment]
There are similar messages... here's my hack with a "fancy" way to load the function group ;-)

```abap
    " Auto-confirm certain messages (requires SAP Note 1609940)
    PERFORM dummy IN PROGRAM saplstrd IF FOUND.  "load function group STRD
    ASSIGN ('(SAPLSTRD)GT_CONFIRMED_MESSAGES') TO <lt_msg>.
    IF sy-subrc = 0.
      " Object can only be created in package of namespace
      ls_msg-id = 'TR'.
      ls_msg-no = '007'.
      COLLECT ls_msg INTO <lt_msg>.
      " Original system set to "SAP"
      ls_msg-id = 'TR'.
      ls_msg-no = '013'.
      COLLECT ls_msg INTO <lt_msg>.
      " Make repairs in foreign namespaces only if they are urgent
      ls_msg-id = 'TR'.
      ls_msg-no = '852'.
      COLLECT ls_msg INTO <lt_msg>.
      " Make repairs in foreign namespaces only if they are urgent
      ls_msg-id = 'TK'.
      ls_msg-no = '016'.
      COLLECT ls_msg INTO <lt_msg>.
    ENDIF.
```

**Comment by thorsten-wolf-neptune** [comment]
niiiiiiiiice. that hack is better not calling any func module. did you solve this in a fork or how/where did you implement it @mbtools ? In anycase good to cover as many messages as might come there.

**Comment by mbtools** [comment]
It's part of the [installer](https://github.com/Marc-Bernard-Tools/MBT-Package-Manager/blob/main/src/inst/%23mbtools%23mbt_installer.prog.abap#L39616) that comes with MBT Package Manager

**Comment by mbtools** [comment]
closed by #6317
